tap_ringcentral/client.py

```diff
@@ -2,52 +2,76 @@
 import base64
 import backoff
 import time
-
+import json
 import singer
 import singer.metrics
 
+from requests.auth import HTTPBasicAuth
+
 LOGGER = singer.get_logger()  # noqa
 
 
 class APIException(Exception):
     pass
 
 
+class AuthFailedException(APIException):
+    """Raised when authentication fails irrecoverably"""
+    pass
+
+
 class RingCentralClient:
 
     MAX_TRIES = 7
 
-    def __init__(self, config):
+    def __init__(self, config, config_path):
         self.config = config
+        self.config_path = config_path
         self.base_url = self.config.get('api_url')
 
         self.refresh_token, self.access_token = self.get_authorization()
 
+    def write_config(self, data):
+        self.config.update(data)
+        with open(self.config_path, "w") as tap_config:
+            json.dump(self.config, tap_config, indent=2)
+
     def get_authorization(self):
         client_id = self.config.get('client_id')
         client_secret = self.config.get('client_secret')
-        basic_auth = base64.b64encode("{}:{}".format(client_id, client_secret).encode())
+        auth = HTTPBasicAuth(client_id, client_secret)
 
-        body = {
-            'username': self.config.get('username'),
-            'password': self.config.get('password'),
-            'grant_type': 'password'
+        payload = {
+            'refresh_token': self.config.get('refresh_token'),
+            'grant_type': 'refresh_token'
         }
 
         headers = {
-            'Authorization': 'Basic {}'.format(basic_auth.decode()),
             'Content-Type': 'application/x-www-form-urlencoded'
         }
 
         response = requests.request(
             'POST',
             '{}/restapi/oauth/token'.format(self.base_url),
+            auth=auth,
             headers=headers,
-            data=body)
+            data=payload)
+
+        if response.status_code == 400 and 'invalid_grant' in response.text:
+            LOGGER.error(
+                'Authentication failed: refresh token is probably expired, invalid, revoked or malformed. '
+                'Re-authenticate to obtain a new refresh token and restore access.'
+            )
+            raise AuthFailedException('Refresh token expired or invalid – auth failed')
 
         response.raise_for_status()
-        json = response.json()
-        return json['refresh_token'], json['access_token']
+        data = response.json()
+
+        # Rotate the refresh token if it’s due to expire within the next 24 hours (86400 seconds)
+        if data['refresh_token_expires_in'] <= 86400:
+            self.write_config({'refresh_token': data['refresh_token']})
+
+        return data['refresh_token'], data['access_token']
```

Comment:
1. The RingCentral API expects that after you refresh, you must start using the new refresh token immediately.
3. You write the new refresh token into config only if it's expiring soon. This may not be correct, because:

- If you do not update the stored refresh token immediately after every refresh, your next run may use a now-invalid old token (if >10s or if the access token has been used).

- The "expires in 24h" logic is probably not needed. You should always persist the new refresh token after a successful refresh.